cluster-app-values-schema/README.md

```diff
@@ -0,0 +1,199 @@
+```
+
+Best practices:
+
+- Do not repeat the property name or title in the description.
+- Write descriptions between 50 and 200 characters long.
+- Use simple language.
+- Use no formatting syntax, nor hyperlinks.
+- Use no line brakes.
+- Use sentence case and punctuation.
+
+### R6: Properties should provide examples {#r6}
+
+Each property SHOULD provide one or more examples using the `example` keyword.
+
+Example:
+
+```json
+{
+  ...
+  "properties": {
+    "name": {
+        "type": "string",
+        "examples": ["devel001"],
+        ...
+    }
+  }
+}
+```
+
+Best practices:
+
+- Ideally, provide one example.
+- Provide more than one example if you want to highlight that very different values are accepted.
+- Provide no more than five examples
+
+TODO
+
+- We could decide to use the examples for testing purposes, replacing ci-values.yaml and the likes. In that case, we should make it a MUST requirement.
+
+### R7: Constrain values as much as possible {#r7}
+
+Property schema SHOULD explicitly restrict values as much as possible.
+
+There are several ways this requirement can/has to be fulfilled, depending on the property type and the use case.
+
+Properties of type `string`:
+
+- Use of the keywords `minLength` and `maxLength` to specify the valid length.
+- If applicable, use the `pattern` keyword to specify a regular expression for validation. This is useful for example to restrict the value to ASCII characters, or to lowercase etc.
+- Use the `format` keyword to restrict the property to a common string format, like a URL or an email address. Watch out to use only formats which are available in the JSON schema dialect used (see also R1).
+
+Numeric properties (type `number`, `integer`):
+
+- Restrict the values range using `minimum` or `exclusiveMaximum` and `maximum` or `exclusiveMaximum`.
+- Use `multipleOf` if the value must be a multiple of a certain number.
+
+Example:
+
+```json
+"diskSizeInGb": {
+  "type": "integer",
+  "minimum": 10,
+  "maximum": 200,
+  "multipleOf": 10
+}
+```
+
+### R8: Required properties must be marked as such {#r8}
+
+If a property is required to be set, it MUST be included in the `required` keyword.
+
+In terms of a lifecycle, this means that the property must be defined when submitting the values for validation or for cluster creation.
+
+
+## TODO
+
+I haven't gotten to these yet, or I'm not sure about them.
+
+- Each property must declare the `type`. Not sure this is required by JSON schema.
+
+- Use of the `default` keyword. As per JSON Schema documentation, it is for annotation only. It is not meant to automatically fill in missing values for validation.
+
+- How to specify whether a property can be modified or not. `"readOnly": true` mit be the right one for that.
+
+- Use of the `deprecated` keyword to phase out properties. Documentation says: "The deprecated keyword is a boolean that indicates that the instance value the keyword applies to should not be used and may be removed in the future."
```

Comment:
I would definitely add this as mandatory. Deprecation notice as the most minimal form of versioning makes contracts for common schemas way easier. Not just removing things but first deprecating it definitely builds important trust into the schemas

Reply:
In the meantime I learned that Helm does not make use of the `deprecated` property of JSON schema at all. There is a little risk that it might mean something different at some point in the Helm world than in the Giant Swarm world, once they start making use of it. Should that worry us?

Context: https://github.com/helm/helm/issues/10732

Reply:
I don't think there is that many use cases for that word besides actually deprecating things, everything else would be expectation breaking. So I wouldn't worry too much - or at least not yet :D.